### ISSUE TITLE
Payment registered in Stripe when OFN declared a problem

### ISSUE THREAD

**Issue by Cecilia-Hn** [openfoodfoundation/openfoodnetwork#7692]
## Description

We've got some order R453734851 for hub ID 844 with this behavior : 

1 - An order is done through OFN
2 - Payment done through Stripe 
3 - Stripe indicates that the amount has been paid
4 - OFN declared a problem with the payment 

Its a big problem for hub and their clients because the order status is invalid so the hub doesnt prepare the order and the client come to collect an order that wasn't prepared.

## Expected Behavior

See the order on the OFN side with the "completed" status

## Actual Behaviour

Order is claimed as incomplete by OFN

## Steps to Reproduce

we don't know :(

## Animated Gif/Screenshot

OFN : 

![image](https://user-images.githubusercontent.com/6525576/119694674-ba29e880-be4d-11eb-8c7e-1ebd92d7ff3e.png)

Stripe : 

![image](https://user-images.githubusercontent.com/6525576/119695749-c95d6600-be4e-11eb-8b24-821a35fd182f.png)

## Workaround
None

## Severity
bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Version used: v3.7.2
* Browser name and version: All
* Operating System and version (desktop or mobile): Both

**Comment by RachL** [comment]
@Matt-Yorkley yes I think there is a really good chance that the bug is fixed. But how should we deal with past orders? Do we have an easy way to fetch all the orders that have had the problem so we can reimburse them? Like are all orders at payment states or should we consider other states as well?

**Comment by lbwright22** [comment]
Another example happened today R267366584 31st May 2021 @Matt-Yorkley

**Comment by filipefurtad0** [comment]
There is a related scenario (stock reduction between order placement and payment is authorization) which is a follow-up from the previous issue, and currently being fixed - see #7665. Maybe it accounts for re-ocurrence @RachL @lbwright22?

**Comment by Cecilia-Hn** [comment]
New case happened yesterday, order : **R315312718**
(the payment is registered on Stipe, nothing on OFN)

**Comment by RachL** [comment]
@Matt-Yorkley FR was rolled back? On slack you've mentioned Aus, US, and Katuma only if I'm not mistaken?

**Comment by Matt-Yorkley** [comment]
Ah! My mistake! Yes, it was Katuma, not France. :man_facepalming:

**Comment by Matt-Yorkley** [comment]
Ah ok... so with **R315312718** the order never actually completed the checkout process? There was an error and the checkout didn't proceed to the order completion page? Is that correct @Cecilia-Hn ?

**Comment by Cecilia-Hn** [comment]
There were no order in OFN and no order confirmation as well

**Comment by Matt-Yorkley** [comment]
**Dev notes:**

There's a lot of moving parts here! There were no relevant Bugsnag errors for that order, which is very strange. It looks like it's possible in some rare cases to process a payment but for the order to subsequently not compete (for some other reason, like stock running out), which creates this situation where the payment has been taken but the order was not completed.

This is due to the order we do things in:
- `Order#process_payments!` gets called in the `before_transition to: :complete` hook
- `Order#finalize!` gets called in the `after_transition to: :complete` hook, and this is where a) stock is reduced and the b) order is marked as completed

If something unexpected happens between those two steps, the payment will be complete but the order will not be :point_up: 

There are a couple of things we can do here:
- add a check for this specific situation during checkout processing, and add an explicit Bugnsag call if it's encountered, to increase visibility of this potential issue and aid debugging
- automatically void the stripe payment if we encounter this issue where the order completion failed (for whatever reason) but the payment had already processed (captured)

**Comment by Matt-Yorkley** [comment]
Alright, I did a bit of a live-debugging deep-dive today and mapped out the code paths for the different flows where additional authorization is required or not required. Also tried to map out edge cases that could cause issues, and found a couple. The flows look a bit like this:

With no additional authorization required:

![Screenshot from 2021-06-09 14-13-50](https://user-images.githubusercontent.com/9029026/121419023-d0669700-c96b-11eb-944b-4e942024f4f9.png)

With additional authorization (the `redirect!` bit is where the user leaves OFN and goes to Stripe):

![Screenshot from 2021-06-09 14-56-14](https://user-images.githubusercontent.com/9029026/121419166-f724cd80-c96b-11eb-887a-1ab03c9a65cd.png)

**Comment by Matt-Yorkley** [comment]
Edge case issues:

**When the stock runs out between being redirected to Stripe and being redirected back to OFN**
- The user gets bounced to the cart page and shown the normal feedback that says items in the cart have gone out of stock, which is good. What they don't get is feedback letting them know what happened with the payment, which is not good. 
- The payment object does _not_ get captured (which is good), but it stays there on the order (we should probably take an action here like marking it as failed instead)
- If the user edit their cart at this point and tries again (and is successful) we still have that previous payment sitting around in pending state, alongside the new completed payment.
- The order's state stays as "payment". We should probably use the "restart checkout" logic here, where we essentially reset the order to cart state.

**When authorization fails the first time, and is then retried**
- This can result in multiple payments on the order in different states (like one failed and one complete). The code doesn't always handle this well.
- There seems to be some issues around lack of explicit ordering in the code that fetches the "last" payment; the returned payment is not always the one we expect!

**Comment by Cecilia-Hn** [comment]
Hi all, 
There is new report from our client, 
Order : R475133320 
Date : 9.06
OFN hub Client : Delphine
Final customer : Sophie - sophie.schoenhoefer@gmail.com 

The 9th june, Delphine is connected to stripe, she sees Sophie's payment uncaptured
Delphine went on OFN and saw Sophie's order as unfinalised 
Delphine went back on stripe for catupring Sophie's payment but she didn't need to do any action, the payment was captured magically 
Delphine went back on OFN and Sophie's order was well finalised

**Comment by Cecilia-Hn** [comment]
Hi ! 
We've got new informations, hope it helps cause our customer is very annoyed by this :/

An OC was open until 11.59 PM on 16th june, 
three payment were on pending status for this same OC until 11.20 AM this morning,
right after that ou producer connected at her Stripe account at 11.15 AM. 
The three orders concerned about this weird payment delay are : 

- R388335642
- R062344327
- R214508066

Does someone have an idea of capture payment delay ?
@Matt-Yorkley

**Comment by Cecilia-Hn** [comment]
A new case : 

A customer has done an order R864750185 this morning. 
Then this customer realized that he did not received the confirmation order by email, so he contact the hub, then, the hub has connect to his stripe account and saw that the payment was not captured by stripe and "magically", the client received the confirmation order by mail and the payment was captured. 

Have we got a clue ?

@Matt-Yorkley

**Comment by RachL** [comment]
@Cecilia-Hn have you check on sendgrid when the confirmation email was sent? If it was right away we can rule out any magic tricks :)

**Comment by Cecilia-Hn** [comment]
Juste received those informations from Stripe : 

<img width="830" alt="Capture d’écran 2021-06-29 à 09 21 55" src="https://user-images.githubusercontent.com/67476305/123755005-c5908980-d8bb-11eb-8338-4b8ab0921fac.png">

And the confirmation mail informations : 

<img width="574" alt="Capture d’écran 2021-06-29 à 09 25 36" src="https://user-images.githubusercontent.com/67476305/123755293-09838e80-d8bc-11eb-8ea7-e7ce3850b3c6.png">

**Comment by Cecilia-Hn** [comment]
About Stripe for R864750185 , we've got these information : 
<img width="830" alt="Capture d’écran 2021-06-29 à 09 21 55" src="https://user-images.githubusercontent.com/67476305/123755896-ac3c0d00-d8bc-11eb-9a0a-dd70b4859784.png">


About Sendgrid : 

<img width="574" alt="Capture d’écran 2021-06-29 à 09 25 36" src="https://user-images.githubusercontent.com/67476305/123755949-b827cf00-d8bc-11eb-9186-6f42d201f0ef.png">

**Comment by RachL** [comment]
Nice! So confirmation email are sent only when there is a capture on Stripe side. It reminds me of https://github.com/openfoodfoundation/openfoodnetwork/issues/7417 where a guy from Stripe was suggesting we could remove this pre-capture: 

_Currently CoopCircuits manages this pre-authorization with capture_method = manual on all your payments. If this parameter is passed to automatic, or is not passed, your payments will no longer use the pre-authorization feature. You can ask the CoopCircuit team if this integration is available._

Maybe a clue?

**Comment by lbwright22** [comment]
An occurrence very similar to this happened in UK yesterday: 
Order ID R744745111
Customer received this message on checkout:
![image](https://user-images.githubusercontent.com/45104035/123767695-93812680-d8bf-11eb-8819-5b8532fc511f.png)

and assumed order had not been placed.  However, order confirmation email was sent and money from customer's Stripe account was taken at the time.

Bugsnag is here: https://app.bugsnag.com/yaycode/open-food-network-js-uk/errors/5f587e8727d37a0017aabc06

**Comment by andrewpbrett** [comment]
I took a look through this issue, I think there's potentially several different things going on, some of which are likely fixed in v3.8.4 and/or by @Matt-Yorkley's PR https://github.com/openfoodfoundation/openfoodnetwork/pull/7779, which is ready for testing.

However, these last few comments sound really odd, so I want to clarify what we're saying here @Cecilia-Hn - it sounds like these are the steps to reproduce:

- Order is placed via shopfront checkout
- Customer sees an error message, does not receive confirmation email
- Hub manager logs into OFN backend some time later and sees payment as Pending in OFN
- Hub manager logs into their Stripe account **and does nothing else**, or maybe looks at the payment but does **not** ever click the capture button?? <-- this is the part that seems really odd
- Payment is captured in Stripe, and OFN order is finalized?

Is that right?

**Comment by Cecilia-Hn** [comment]
Hi Andrew and thanks a lot.
I know it's very odd. 

Your steps seems correct except that the customer does not saw error message. 
_

- Order is placed via shopfront checkout
- Customer does not receive confirmation email
- Hub manager logs into OFN backend some time later and sees payment as Pending in OFN
- Hub manager logs into their Stripe account and does nothing else, or maybe looks at the payment but does not ever click the capture button
- Payment is captured in Stripe, and OFN order is finalized

_

**Comment by andrewpbrett** [comment]
Thanks for the update @Cecilia-Hn! Super bizarre.... I'm a bit stuck as to an explanation for this, so if Stripe does reply with any more information please post it here!

**Comment by RachL** [comment]
@andrewpbrett @filipefurtad0 so Stripe explained that after authorization, our set up requires capture, which (without hub managers doing anything) happens automatically after 7 days. So the money is "on hold" at the shopper side during all this time.

I don't understand why we only see this problem now...

Is it because shoppers get re-directed to the authorization page, and could close the window before coming back to OFN - thus not knowing the order had a problem?

**Comment by filipefurtad0** [comment]
That's really useful info @RachL ,

> I don't understand why we only see this problem now...

I don't understand that either...

But at least now it is clear why `uncaptured` payments are captured automatically. Tbh, this puzzled me quite a bit, as [until recently](https://github.com/openfoodfoundation/openfoodnetwork/pull/7497) this was the case for all subscription payments (orders were complete, despite the `uncaptured` payment).

So, since last release, we have #7779, which is warning the user something went wrong on checkout and displaying a cancelled payment message (which is currently left as `uncaptured` - #7922) - this might help identify more these occurrences (should they happen).

**Comment by RachL** [comment]
@Cecilia-Hn so far our only clue is what is now describe here: https://github.com/openfoodfoundation/openfoodnetwork/issues/7922 (orders where dealing with products going out of stock during the authorization process).

I'm closing here. If we have other requests coming in which are not linked to a stock problem, we can reopen.

**Comment by Cecilia-Hn** [comment]
Hi, 

I think we may need to re open this issue. 
We've got a new case from R207347030 order : the payment through Stripe is captured when OFN Payment is still uncaptured. 

What do you think ? 

All informations bellow : 


<img width="1084" alt="Capture d’écran 2021-08-20 à 16 09 56" src="https://user-images.githubusercontent.com/67476305/130246248-848a4251-abee-4e6e-8c67-5e14e099581f.png">
<img width="989" alt="Capture d’écran 2021-08-20 à 16 10 10" src="https://user-images.githubusercontent.com/67476305/130246279-e12b9325-c96b-433e-940f-cac70d055432.png">
<img width="1104" alt="Capture d’écran 2021-08-20 à 16 10 25" src="https://user-images.githubusercontent.com/67476305/130246312-14b73af4-1ed0-42f6-858e-14b0a5fc36d3.png">

**Comment by Cecilia-Hn** [comment]
@filipefurtad0 Thanks !
Yes, it was a shopfront order.

**Comment by lin-d-hop** [comment]
We are still seeing this in UK as well. The pain it is causing is high! And I am not sure all the occurrences are actually being reported.

**Comment by lin-d-hop** [comment]
Always through the shopfront for UK experience.

**Comment by filipefurtad0** [comment]
The payment (`id=179982`) for that order is in state `failed` has a `cvv_response_message` like`source_redirect_slug` (replaced some customer info with `<removed_string>` below):

`https://hooks.stripe.com/redirect/authenticate/src_1JPWCVKNkMQwqy0M6jONUjaB?client_secret=src_client_secret_<removed_string>source_redirect_slug=live_<removed_string>
`

There are 52 such payments in FR-prod:

```
openfoodnetwork=> SELECT COUNT(*) FROM spree_payments WHERE cvv_response_message LIKE '%source_redirect_slug%';
 count 
-------
    52
(1 row)

```

And 4 of them UK-prod:

```
openfoodnetwork=> SELECT COUNT(*) FROM spree_payments WHERE cvv_response_message LIKE '%source_redirect_slug%';
 count 
-------
     4
(1 row)
```

Not sure they all relate to the same scenario.

We've seen a similar message before, on a different context:
https://github.com/openfoodfoundation/openfoodnetwork/issues/7687

**Comment by filipefurtad0** [comment]
If the order cycle closes during checkout on the shopfront, and while payment is being processed, then:
- the customer will be redirected to the shopfront, or homepage - as expected, depending if there are more OC open
- the order will remain in a state = `payment`; no order confirmations arrive
- the payment remains authorized and `uncaptured`, on Stripe

![image](https://user-images.githubusercontent.com/49817236/130273307-8a07cf84-9d43-4160-99c2-db60bd0bb2fd.png)

![Peek 2021-08-20 18-40](https://user-images.githubusercontent.com/49817236/130273376-17e33f37-604a-49c4-b36b-c18bee7ca4ee.gif)

This seems to be different than the scenario above though... In this case, the order does not have any payments:

![image](https://user-images.githubusercontent.com/49817236/130273768-d1175740-7f0e-4cee-a873-820c0e3d0ec6.png)

So there might be more scenarios which account for this bug.

**Comment by andrewpbrett** [comment]
Thanks for the update @Cecilia-Hn - definitely correct to reopen this. That screenshot you provided was super helpful. With the timestamp I was able to look through the logs on FR and find the request that we saw after the customer authorized and was redirected to OFN.

The next line in the logs is:

`Redirected by /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2021-08-10-144554/app/controllers/application_controller.rb:127:in 'require_distributor_chosen'`

So we're not marking the payment as complete in OFN because of that redirect. I think @filipefurtad0 is on the right track here, but I need to dig a little more to figure out why `current_distributor` might be false, it looks like there's a few different paths to go down. 

In any case, though, it might be a good idea to move the handling of the redirect from Stripe to its own controller action entirely.

**Comment by lin-d-hop** [comment]
Multiple reports a week of this. This is one of the worst bugs we've had and its hung around all of August. There's no one to work on it but it is an S1.

**Comment by RachL** [comment]
@andrewpbrett stripe went back to me and asked that a developer contact them directly. They are happy to have a look at our integration. They are available through their chat system or with https://support.stripe.com/contact/email?topic=api_integration

You will need to use this link while being loogedin FR's account (this way they answer faster as they see we are customers). You can do it with your personal access you have on FR account. Otherwise, if you log in with the credentials on bitwarden there is a 2FA on my phone, so I need to be around^^

**Comment by andrewpbrett** [comment]
For the specific payment most recently posted by @Cecilia-Hn, the problem was that we were [redirecting to the home page](https://github.com/openfoodfoundation/openfoodnetwork/issues/7692#issuecomment-902999901). 

As @Matt-Yorkley [theorized](https://github.com/openfoodfoundation/openfoodnetwork/pull/8089#issuecomment-904998473), this may be due to the person being logged in on the non-www domain and then being redirected by the payment provider to the www domain (or vice versa). 

To test this theory, I started a cart session and entered a card that requires SCA redirection. When I got to the SCA auth page I copied the URL and visited it in a new private window, to simulate the situation where the customer is authorizing the SCA payment but not logged in on the domain that they're about to be redirected to. Sure enough, I got: 

`Redirected by [...]/application_controller.rb:127:in 'require_distributor_chosen'`

So I believe the specific, most latest scenario described by this issue will be resolved by #8090, which is already merged and will be in the next release. 

There are two pieces that are still outstanding:

It would be good to look into whether the 56 payments that @filipefurtad0 found and see if we can verify if they were all related to this same scenario (non-www vs www). I'll take some time to do that now. If it starts to take longer than an hour or two, I'd say we should move it to its own issue and close this one.

**Comment by andrewpbrett** [comment]
The four payments in the UK that @filipefurtad0 found are all from late July/early August. So @lin-d-hop's report that this is still happening multiple times per week looks like a separate issue. #8090 will prevent the scenario that @Cecilia-Hn most recently mentioned, but probably not whatever is still happening in the UK. 

@lin-d-hop are you able to provide any order numbers of examples that you've seen recently to help narrow this down?

**Comment by Matt-Yorkley** [comment]
Seems like this case where the order cycle closes just before the user gets redirected back is still a good candidate for improvement, right? https://github.com/openfoodfoundation/openfoodnetwork/issues/7692#issuecomment-902857548

**Comment by andrewpbrett** [comment]
> Seems like this case where the order cycle closes just before the user gets redirected back is still a good candidate for improvement, right?

I think so, yes; but I think it's unlikely that that's the root cause of these payment issues. If the order cycle closes while the customer is performing the SCA auth, the customer gets a clear message indicating what happened. They're also redirected by the `check_order_cycle_expiry` method; the payment that @Cecilia-Hn mentions was redirected by `require_distributor_chosen`. 

I looked at the logs in the UK, and there's nothing that matches the entries you would expect for the order cycle closing scenario: there are no redirects by the `check_order_cycle_expiry` method while attempting to process a payment intent in any recent logs (I went back to August 18). 

So I believe:
- #8090 will, once deployed, resolve one way that this was happening
- The scenario where the order cycle closes while the customer is authorizing the payment needs a separate issue (I'll create one). As far as I can tell, this scenario *has not happened* recently in the UK. 
- There's still potentially a third scenario that we're not accounting for. I think it also needs a separate issue (I can also create this one unless @lin-d-hop has additional info on them) I looked for any payments in the UK that are currently in `state = requires_authorization` and there are 16; these are probably a good place to start investigating. These are their id's and `updated_at`:

```
[["2021-05-16", 374727], ["2021-05-25", 377352], ["2021-05-26", 377522], ["2021-07-30", 393002], ["2021-07-31", 393154], ["2021-07-31", 393157], ["2021-08-02", 393404], ["2021-08-05", 394307], ["2021-08-07", 394667], ["2021-08-08", 394957], ["2021-08-09", 395301], ["2021-08-09", 395306], ["2021-08-19", 397252], ["2021-08-23", 397055], ["2021-08-25", 398568], ["2021-09-02", 400105]]
```
